src/main/java/com/ibm/cloud/sdk/core/security/IamToken.java

```diff
@@ -52,27 +54,46 @@ public Long getExpiration() {
   }
 
   /**
-   * Check if currently stored access token is valid.
+   * Returns true iff currently stored access token should be refreshed.
    *
-   * Using a buffer to prevent the edge case of the
-   * token expiring before the request could be made.
+   * This method uses a buffer to prevent the edge case of the
+   * token expiring before the request could be made, so this method will return true
+   * if the current time is within that buffer.
    *
    * The buffer will be a fraction of the total TTL. Using 80%.
    *
-   * @return true iff the current access token is valid and not expired
+   * This method also updates the expiration time if it determines the token needs refreshed to prevent other
+   * threads from making multiple refresh calls.
+   *
+   * @return true iff the current access token is invalid or past the expiration buffer
    */
   @Override
-  public boolean isTokenValid() {
-    if (getAccessToken() == null || getExpiresIn() == null || getExpiration() == null) {
-      return false;
+  public synchronized boolean needsRefresh() {
+    if (this.refreshTime == null && getExpiresIn() != null && getExpiration() != null) {
+      Double fractionOfTimeToLive = 0.8;
+      Long timeToLive = getExpiresIn();
+      Long expirationTime = getExpiration();
+      this.refreshTime = expirationTime - (long) (timeToLive * (1.0 - fractionOfTimeToLive));
+    }
+
+    if (this.refreshTime != null && Clock.getCurrentTimeInSeconds() > this.refreshTime) {
+      // Advance refresh time by one minute.
+      this.refreshTime += 60;
+
+      return true;
     }
 
-    Double fractionOfTimeToLive = 0.8;
-    Long timeToLive = getExpiresIn();
-    Long expirationTime = getExpiration();
-    Double refreshTime = expirationTime - (timeToLive * (1.0 - fractionOfTimeToLive));
-    Double currentTime = Math.floor(System.currentTimeMillis() / 1000);
+    return false;
+  }
 
-    return currentTime < refreshTime;
+  /**
+   * Check if the currently stored access token is valid. This is different from the needsRefresh method in that it
+   * uses the actual TTL to calculate the expiration, rather than just a fraction.
+   *
+   * @return true iff is the current access token is not expired
+   */
+  @Override
+  public boolean isTokenValid() {
+    return Clock.getCurrentTimeInSeconds() < this.expiration;
```

Comment:
It's a little strange that we use `this.expiration` here, rather than `getExpiration()`, as we do in `needsRefresh`, and that we don't check for null here where we did in `needsRefresh`.  So could this property be null or not?  We seem to be schizophrenic on this.

For consistency, I think this line should be:
```
return this.expiration == null || Clock.getCurrentTimeInSeconds() < this.expiration;
```

So a null expiration time means "never expires".  I actually don't think this will happen, but again, for consistency, I think we need to handle it here.